### ISSUE TITLE
Bokeh not purging data between calls? Ballooning file sizes!

### ISSUE THREAD

**Issue by guillochon** [bokeh/bokeh#3671]
Hi, I have a loop that's producing HTML via Bokeh, and I've noticed something very strange where the HTML files are constantly growing in size, despite the fact that the plotted data is always about the same size (hundreds of kB to a few MB per dataset). Eventually the files are 500MB in size for ~1MB datasets!

What's strange is that the generated HTML still visually produces the right plot for each dataset, it's not like I can see all the previous datasets up to then. Here's the first plot generated in the loop: https://sne.space/sne/SN1993ac.html

...and here's the plot generated 20th in the loop (which is already 100x bigger in file size than the first plot!), note that this plot actually has _less data_ than the first example: https://sne.space/sne/SN1996ac.html

I have verified that if I generate that 20th plot first that its filesize is normal. Below is my code. Does anyone have any idea what's going on?

``` Python
if spectraavail and dohtml and args.writehtml:
        spectrumwave = []
        spectrumflux = []
        spectrumerrs = []
        for spectrum in catalog[entry]['spectra']:
            specrange = range(len(spectrum['data']))
            spectrumwave.append([float(spectrum['data'][x][0]) for x in specrange])
            spectrumflux.append([float(spectrum['data'][x][1]) for x in specrange])
            if 'errorunit' in spectrum:
                spectrumerrs.append([float(spectrum['data'][x][2]) for x in specrange])

        y_height = 0.
        y_offsets = [0. for x in range(len(spectrumwave))]
        for i in reversed(range(len(spectrumwave))):
            y_offsets[i] = y_height
            ydiff = max(spectrumflux[i]) - min(spectrumflux[i])
            spectrumflux[i] = [j + y_height for j in spectrumflux[i]]
            y_height += ydiff

        maxsw = max(map(max, spectrumwave))
        minsw = min(map(min, spectrumwave))
        maxfl = max(map(max, spectrumflux))
        minfl = min(map(min, spectrumflux))
        maxfldiff = max(map(operator.sub, list(map(max, spectrumflux)), list(map(min, spectrumflux))))
        x_buffer = 0.0 #0.1*(maxsw - minsw)
        x_range = [-x_buffer + minsw, x_buffer + maxsw]
        y_buffer = 0.1*maxfldiff
        y_range = [-y_buffer + minfl, y_buffer + maxfl]

        for f, flux in enumerate(spectrumflux):
            spectrumflux[f] = [x - y_offsets[f] for x in flux]

        tt2 = [  
                ("λ", "@x{1.1}"),
                ("Flux", "@y0"),
                ("Epoch (" + spectrum['timeunit'] + ")", "@epoch{1.11}"),
                ("Source", "@src")
             ]
        hover2 = HoverTool(tooltips = tt2)

        p2 = Figure(title='Spectra for ' + eventname, x_axis_label=label_format('Wavelength (' + catalog[entry]['spectra'][0]['waveunit'] + ')'),
            y_axis_label=label_format('Flux (' + catalog[entry]['spectra'][0]['fluxunit'] + ')' + ' + offset'
            if (len(catalog[entry]['spectra']) > 1) else ''), x_range = x_range, tools = tools, 
            y_range = y_range)
        p2.add_tools(hover2)

        colors = brewer["Spectral"]
        colors = colors[min(max(min(colors.keys()),len(spectrumwave)),max(colors.keys()))]
        sources = []
        for i in range(len(spectrumwave)):
            sources.append(ColumnDataSource(
                data = dict(
                    x0 = spectrumwave[i],
                    y0 = spectrumflux[i],
                    x = spectrumwave[i],
                    y = [y_offsets[i] + j for j in spectrumflux[i]],
                    yoff = [y_offsets[i]],
                    binsize = [1.0],
                    spacing = [1.0],
                    src = [catalog[entry]['spectra'][i]['source'] for j in spectrumflux[i]],
                    epoch = [catalog[entry]['spectra'][i]['time'] for j in spectrumflux[i]]
                )
            ))
            p2.line('x', 'y', source=sources[i], line_color=str(colors[i % len(colors)]), line_width=2)

        sdicts = dict(zip(['s'+str(x) for x in range(len(sources))], sources))
        callback = CustomJS(args=sdicts, code="""
            for (s = 0; s < """ + str(len(sources)) + """; s++) {
                var data = eval('s'+s).get('data');
                if (cb_obj.get('title') == 'Spacing') {
                    data['spacing'][0] = cb_obj.get('value');
                } else {
                    data['binsize'][0] = cb_obj.get('value');
                }
                var f = data['binsize'][0]
                var space = data['spacing'][0]
                var x0 = data['x0'];
                var y0 = data['y0'];
                var dx0 = x0[1] - x0[0];
                var yoff = space*data['yoff'][0];
                data['x'] = [x0[0] - 0.5*Math.max(0., f - dx0)];
                data['y'] = [y0[0] + yoff];
                var xaccum = 0.;
                var yaccum = 0.;
                for (i = 0; i < x0.length; i++) {
                    var dx;
                    if (i == 0) {
                        dx = x0[i+1] - x0[i];
                    } else {
                        dx = x0[i] - x0[i-1];
                    }
                    xaccum += dx;
                    yaccum += y0[i]*dx;
                    if (xaccum >= f) {
                        data['x'].push(data['x'][data['x'].length-1] + xaccum);
                        data['y'].push(yaccum/xaccum + yoff);
                        xaccum = 0.;
                        yaccum = 0.;
                    }
                }
                eval('s'+s).trigger('change');
            }
        """)

        binslider = Slider(start=0, end=20, value=1, step=0.5, title=label_format("Bin size (Angstrom)"), callback=callback)
        spacingslider = Slider(start=0, end=2, value=1, step=0.02, title=label_format("Spacing"), callback=callback)

    #if (photoavail or spectraavail) and dohtml and args.writehtml:
    if (photoavail and spectraavail) and dohtml and args.writehtml:
        if photoavail and spectraavail:
            p = HBox(vplot(hplot(p1,p2,vform(binslider,spacingslider))))
        elif photoavail:
            p = p1
        else:
            p = p2

        html = file_html(p, CDN, eventname)
        returnlink = r'<br><a href="https://sne.space"><< Return to supernova catalog</a>'
        repfolder = get_rep_folder(catalog[entry])
        html = re.sub(r'(\<\/body\>)', r'<a href="https://cdn.rawgit.com/astrotransients/' + repfolder + '/master/' + eventname + r'.json" download>Download datafile</a><br><br>\n\1', html)
        if len(catalog[entry]['sources']):
            html = re.sub(r'(\<\/body\>)', r'<em>Sources of data:</em><br><table><tr><th width=30px>ID</th><th>Source</th></tr>\n\1', html)
            for source in catalog[entry]['sources']:
                html = re.sub(r'(\<\/body\>)', r'<tr><td>' + source['alias'] +
                r'</td><td>' + source['name'].encode('ascii', 'xmlcharrefreplace').decode("utf-8") +
                r'</td></tr>\n\1', html)
            html = re.sub(r'(\<\/body\>)', r'</table>\n\1', html)
        html = re.sub(r'(\<\/body\>)', returnlink+r'\n\1', html)
        print(outdir + eventname + ".html")
        with open(outdir + eventname + ".html", "w") as f:
            f.write(html)
```

**Comment by guillochon** [comment]
Another note: I checked, and the data for the first *is in the HTML for the 20th plot*. So the data is being accumulated in the loop, and that's why the file sizes are exploding. It appears that the `ColumnDataSource` objects defined in previous iterations of the loop are finding their way into the Bokeh data structure.

**Comment by guillochon** [comment]
OK I think I found a solution in this StackExchange thread: http://stackoverflow.com/questions/24739390/updated-bokeh-to-0-5-0-now-plots-all-previous-versions-of-graph-in-one-window, you have to run `bokeh.plotting.reset_output` at the end of every iteration of the loop.

But it still seems to me that Bokeh shouldn't be so "sticky" in this situation. The data structures get out of control very quickly!

**Comment by bryevdv** [comment]
@guillochon there are definitely trade-offs between use-cases. One of the most important use-cases for us is to make creating deployable apps dead-simple, and to do this we instrumented Bokeh models to record changes automatically. But one result is what you see here, so the trade-off for less code in one case might be more code in another case. That said, it might be possible and reasonable to prune orphaned models during a `show()` (at least by default)  @havocp thoughts?

**Comment by havocp** [comment]
I'm not sure I understand the situation here but I think vplot/hplot/vform are auto-adding to bokeh.io.curdoc and nothing is ever removed from curdoc?

If so a simple fix is make your document explicitly `doc = Document()` on each loop and add the plot to it with add_root and don't use vplot etc just use VBox HBox that make the widgets but don't add them anywhere. Then you have no global state doing unexpected stuff.

If you use the global curdoc you have to clear it when you want to start over.

**Comment by havocp** [comment]
I don't see any models being orphaned here, orphaned models wouldn't go in file_html would they? they are all still in curdoc

**Comment by bryevdv** [comment]
I think you are right

**Comment by havocp** [comment]
I think you don't even have to create a document come to think of it as long as you avoid vplot/hplot/vform that pull in the global doc. file_html should then create a doc each time for you when you pass in models that have no doc.

**Comment by bryevdv** [comment]
We absolutely have to figure out what we want to do long term with `Figure`/`figure`, `hplot`, etc very soon. My suggestion to @fpliger was this:

* `figure` stays but no longer immediately adds anything to the current doc
* `hplot`, `vplot` deprecated
* if the document has no roots  --- *only* if the document has no roots ---  try to "do the right thing" on `show`, i.e. add any `Figures` or layouts to the document

So basically, defer the "autoadd" until the last instant, and only if necessary

**Comment by havocp** [comment]
agreed, the auto-add behavior is a big source of confusion and faqs right now. It's not obvious from the names of these functions that they add stuff to a global curdoc(), and the result is understandable surprise.

**Comment by bryevdv** [comment]
Does the above plan seem reasonable?

**Comment by havocp** [comment]
I haven't thought about it too much... agree with the high-level goal of phasing this stuff out so it isn't confusing everyone. Some half-baked reaction:

 * would show() auto-adding also include say push(), and at the end of a bokeh server script? (essentially, everywhere that we validate() maybe we should also see if we need to autoadd?)
 * agree with deprecating hplot/vplot, but maybe if deprecating them don't also change their behavior, since why break something that's going away; or maybe make them work like the new figure()
 * instead of only auto-adding if no roots in curdoc, maybe only auto-add if the figure is completely orphaned (hasn't been added to any other doc) ... not sure. it could get weird, for example you add some unrelated root to curdoc and all your figures go away.

Another thought I had just now... we don't really need stuff to be in curdoc() in order to output it as a file, because file_html creates a transient document anyway IIRC if none of the models passed to it already have a doc. So if people aren't manually dealing with curdoc() or any Document, we could simply rely on it being ok that stuff isn't in a document? Or maybe with some tweaking there's something to that idea, not sure how it would play out.

**Comment by fpliger** [comment]
> agree with deprecating hplot/vplot, but maybe if deprecating them don't also change their behavior, since why break something that's going away; or maybe make them work like the new figure()

+1

**Comment by anadyn** [comment]
Hi guys, thanks for all the efforts you put into the program.
Anyway, as a regular user I can confirm the confusion regarding the auto-add function, and the resulting ballooning file sizes. The recommended installation on the Bokeh getting started is by Anaconda, and then it is natural to run Spyder. By default in Spyder the same interpreter is used the whole time, which leads to the balloning files.

**Comment by bryevdv** [comment]
@guillochon can you comment if this is still an issue, and if so, provide a complete, minimal runnable test case? I had wanted to check myself but the code above is incomplete. I'm willing to believe that "accumulation" still happens in some use cases (absent `reset_output` or `curdoc().clear()` but it would be good to have a minimal demonstration of this to focus discussion around any possible improvements that might be possible.

**Comment by guillochon** [comment]
@bryevdv As I commented above, using `reset_output` fixed this for me, are you suggesting that `reset_output` should no longer be required?

**Comment by bryevdv** [comment]
I'd like to continue this discussion in case there is a way to improve things so explicit clearing is not always needed, but let's do it in the fresh issue.
